File: 200-number-of-islands/200-number-of-islands.cpp
class Solution {
    int ROW, COL;
public:
    int numIslands(vector<vector<char>>& G) {
        ROW = size(G), COL = size(G[0]);
        int count = 0;
        for (int i = 0;i < ROW;i++) {
            for (int j = 0;j < COL;j++) {
                if (G[i][j] == '1') {
                    count += 1;
                    dfs(G, i, j);
                }
            }
        }
        return count;
    }
private:
    void dfs(vector<vector<char>> &G, int i, int j) {
        if (valid(G, i, j)) {
            G[i][j] = '0';
            // call in all dirs
            dfs(G, i+1, j);
            dfs(G, i-1, j);
            dfs(G, i, j+1);
            dfs(G, i, j-1);
        } else return;
    }
    bool valid(vector<vector<char>> &G, int i, int j) {
        return i >= 0 and
            j >= 0 and
            i < ROW and
            j < COL and
            G[i][j] == '1';
    }
};

File: 200-number-of-islands/200-number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        m, n, cnt = len(grid), len(grid[0]),0
        
        def dfs(i,j):
            if (i < 0 or j < 0 or i >= m or j >= n or grid[i][j] != '1') :
                return
            grid[i][j] = '0'
            # down
            dfs(i+1,j)
            # up
            dfs(i-1,j)
            # left
            dfs(i,j-1)
            # right
            dfs(i,j+1)
        
        for i in range(m):
            for j in range(n):
                if grid[i][j] == '1':
                    cnt += 1
                    dfs(i,j)
        return cnt